File: year2018/day03/python/solution.py
import re

def main():
  print(part_1())
  print(part_2())


def part_1():
  fabric, overlaps = [[0] * 1000 for _ in range(1000)], 0

  for _, x, y, w, h in input:
    for i in range(x, x + w):
      for j in range(y, y + h):
        if fabric[i][j] == 1:
          overlaps += 1

        fabric[i][j] += 1

  return overlaps


def part_2():
  fabric, no_overlap = [[0] * 1000 for _ in range(1000)], set()

  for claim_id, x, y, w, h in input:
    no_overlap.add(claim_id)

    for i in range(x, x + w):
      for j in range(y, y + h):
        if fabric[i][j] == 0:
          fabric[i][j] = claim_id

        else:
          no_overlap.discard(claim_id)
          no_overlap.discard(fabric[i][j])

  return int(",".join(map(str, no_overlap)))


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [list(map(int, re.findall(r'\d+', claim))) for claim in file.readlines()]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2018/day03/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

public class Solution {
  private static List<List<Integer>> CODES;

  public static void main(String[] args) {
    try {
      setup(Files.readString(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static int part1() {
    int[][] fabric = new int[1000][1000];
    int overlaps = 0;
    for (List<Integer> claim : CODES) {
      for (int i = claim.get(1); i < claim.get(1) + claim.get(3); i++) {
        for (int j = claim.get(2); j < claim.get(2) + claim.get(4); j++) {
          if (fabric[i][j]++ == 1) {
            overlaps++;
          }
        }
      }
    }
    return overlaps;
  }

  public static int part2() {
    int[][] fabric = new int[1000][1000];
    Set<Integer> noOverlap = new HashSet<>();
    for (List<Integer> claim : CODES) {
      noOverlap.add(claim.get(0));
      for (int i = claim.get(1); i < claim.get(1) + claim.get(3); i++) {
        for (int j = claim.get(2); j < claim.get(2) + claim.get(4); j++) {
          if (fabric[i][j] == 0) {
            fabric[i][j] = claim.get(0);
          } else {
            noOverlap.remove(claim.get(0));
            noOverlap.remove(fabric[i][j]);
          }
        }
      }
    }
    return noOverlap.isEmpty() ? -1 : noOverlap.iterator().next();
  }

  public static void setup(String input) {
    CODES = Arrays.stream(input.split("\n"))
        .map(claim -> Arrays.stream(claim.split("\\D+"))
        .filter(s -> !s.isEmpty())
        .map(Integer::parseInt)
        .collect(Collectors.toList()))
        .collect(Collectors.toList());
  }
}